clang/lib/Sema/SemaExpr.cpp

```diff
@@ -17154,6 +17156,10 @@ ExprResult Sema::ActOnBlockStmtExpr(SourceLocation CaretLoc,
   BlockScopeInfo *BSI = cast<BlockScopeInfo>(FunctionScopes.back());
   BlockDecl *BD = BSI->TheDecl;
 
+  if (const auto FX = BD->getFunctionEffects()) {
+    CheckAddCallableWithEffects(BD, FX);
+  }
```

Comment:
Not too familiar with blocks candidly, but potentially the same thing I said about lambdas might apply here too.